### PR TITLE
Add/limit action run

### PR DESCRIPTION
**Pull request by AHReccese**
#### Reference Issues/PRs
- limit action runs to just commits on dev & main branches plus PRs to dev branch.
#### What does this implement/fix? Explain your changes.
- Imagine I want to push my local code, but I don't want to trigger actions, maybe I'm not done
- Also right now we have double checks per pull requests, one for the branch and the other one for the PR itself

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - PyMilo exception types added in `pymilo/exceptions/__init__.py`
 - PyMilo exception types added in `pymilo/__init__.py`
 ### Changed
+- action run limit added to `test.yml`
```
Comment (sadrasabouri) [inline CHANGELOG.md:12]:
```
- GitHub actions are limited to the `dev` branch
```

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -3,7 +3,15 @@
...
-on: [push, pull_request]
+on:
+  push:
+    branches:
+      - main
+      - dev
+
+  pull_request:
+    branches:
+      - dev
```
Comment (sadrasabouri) [inline .github/workflows/test.yml:14]: add `main` here as well.
I know that with the current release model we don't PR to `main` but there is no reason to remove it.

**Review by sadrasabouri — CHANGES_REQUESTED** [review]
Good point Amirhossein, that's cruical for our rate limit.
I don't know if it's possible (probably not) but let's see if we can only run GH actions on PRs that have `ready` status (not the draft ones).

**Comment by AHReccese** [comment]
Dear Sadra, I filtered out the `draft` PRs and applied the rest of your feedback.

**Comment by AHReccese** [comment]
I still need to work on this ...

**Comment by AHReccese** [comment]
@sadrasabouri 
Dear Sadra, the most we can do for the draft pull requests is to check the draft field in a pre-job job, I reviewed the GitHub documentations in detail and there isn't a completely correct way to apply filter for bypassing "draft"s in the `on` section of the `test.yml` file.

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - PyMilo exception types added in `pymilo/exceptions/__init__.py`
 - PyMilo exception types added in `pymilo/__init__.py`
 ### Changed
+- GitHub actions are limited to the `dev, main` branches
```
Comment (sadrasabouri) [inline CHANGELOG.md:12]:
```
`dev` and `main`
```

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -3,14 +3,32 @@
...
+  check-draft:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Exit if draft PR
+        if: ${{ github.event.pull_request.draft == true }}
+        run: |
+          echo "This is a draft pull request. Exiting."
+          exit 1
   build:
+    needs: check-draft
```
Comment (sadrasabouri) [inline .github/workflows/test.yml:31]: Can you describe it? Does this works in all cases?
If it has corner cases we can ignore it. It's not necessary.

**Review by sadrasabouri — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml (hunk 4 above)
Comment (AHReccese) [reply .github/workflows/test.yml:31]: @sadrasabouri 

Yes, this approach works in all cases. The key difference with filtering in the `on` section is that, technically, the action still triggers but halts after internal checks, preventing further execution. By applying filtering directly in the `on` command, the action is never triggered in the first place.

Additionally, the code defines a pre-job that runs before other jobs. This pre-job checks whether the pull request is a draft, and if it is, the job exits with a non-zero status, preventing subsequent jobs from running.

**Comment by AHReccese** [comment]
I tested the functionality in `DMeta` and `Reserver` projects.

**Comment by sepandhaghighi** [comment]
@AHReccese 
I'm not a reviewer here, but I have some concerns about this approach. Please keep it open to talk about it in our weekly session.

**Comment by AHReccese** [comment]
> @AHReccese I'm not a reviewer here, but I have some concerns about this approach. Please keep it open to talk about it in our weekly session.

OK sure, we will talk about it.

**Comment by AHReccese** [comment]
@sadrasabouri ready for your review. I removed the job associated with checking draft PRs according to the Sepand's feedback.

**Review by sadrasabouri — APPROVED** [review]
